### PR TITLE
Make _array_types sentinels survive cloudpickle round-trip.

### LINKED ISSUE #390: cloudpickle round-trip of `Foo[Array, "..."]` breaks the anonymous-variadic-dim sentinel
# cloudpickle round-trip of `Foo[Array, "..."]` breaks the anonymous-variadic-dim sentinel

## Summary

cloudpickling a jaxtyping type that uses the `"..."` axis (`Float[Array, "..."]` and friends) yields a type whose variadic-dim entry no longer matches the module-level `_array_types._anonymous_variadic_dim` sentinel. Any subsequent shape check on the unpickled type then trips the `assert type(variadic_dim) is _NamedVariadicDim` in `_array_types._check_dims` (`_array_types.py` L297) and raises `AssertionError`.

Plain `pickle` round-trips fine — the bug surfaces only under `cloudpickle`, which serializes the class structure (and the sentinel reference inside its `dims` tuple) by value.

## Reproducer

`jaxtyping==0.3.9`, `cloudpickle==3.1.2`, `jax==0.10.1`, Python 3.14.5.

```python
import cloudpickle
import jax.numpy as jnp
from jax import Array
from jaxtyping import Float, _array_types

sentinel = _array_types._anonymous_variadic_dim
T = Float[Array, "..."]
assert T.dims[T.index_variadic] is sentinel        # holds before pickle

loaded_T = cloudpickle.loads(cloudpickle.dumps(T))

loaded_sentinel = loaded_T.dims[loaded_T.index_variadic]
print("identity preserved:", loaded_sentinel is sentinel)      # False
print(isinstance(jnp.zeros((2, 3)), loaded_T))                  # AssertionError
```

Output:
```
identity preserved: False
AssertionError
```

(Plain `pickle.dumps(T)` / `pickle.loads(...)` works fine, because stdlib pickle resolves the type by qualified name on the receiving side and reads `_anonymous_variadic_dim` from the live module global.)

## Root cause

`_array_types.py` (L86):

```python
_anonymous_variadic_dim = object()
```

`object()` has no `__reduce__`, so cloudpickle's default reduction creates a fresh instance on unpickle. The new instance is unrelated to the live module global, breaking the `is`-identity check at L294-297:

```python
if variadic_dim is _anonymous_variadic_dim:
    ...
else:
    assert type(variadic_dim) is _NamedVariadicDim   # ← fails
```

The unpickled `variadic_dim` is a plain `object`, neither the live sentinel nor a `_NamedVariadicDim`.

## Proposed fix

Replace the bare `object()` sentinel with a `__reduce__`-backed singleton. On unpickle, `__reduce__` calls the singleton constructor, which returns the cached instance — identity holds.

```python
# _array_types.py, replacing the L86 definition

class _AnonymousVariadicDim:
    """Picklable singleton for the `"..."` axis marker."""

    _instance = None

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
        return cls._instance

    def __reduce__(self):
        return (_AnonymousVariadicDim, ())


_anonymous_variadic_dim = _AnonymousVariadicDim()
```

With this in place, the reproducer above prints:

```
identity preserved: True
True
```

Existing call sites (`is _anonymous_variadic_dim` at L294, L351, L507) continue to work unchanged, because there's still exactly one live instance per process.

## Context

Found while debugging a beartype-claw'd JAX project that cloudpickles DAG-built functions for distribution to worker processes. Currently worked around by monkey-patching `_array_types._anonymous_variadic_dim` at import time: [`pylcm/src/_lcm/jaxtyping_patch.py`](https://github.com/OpenSourceEconomics/pylcm/blob/main/src/_lcm/jaxtyping_patch.py).

Happy to send a PR with the in-tree change above if useful!

[jaxtyping_issue_repro.py](https://github.com/user-attachments/files/28171679/jaxtyping_issue_repro.py)

### PR DESCRIPTION
**Pull request by hmgaudecker**
The three module-level `object()` sentinels in `_array_types` (`_any_dtype`, `_anonymous_dim`, `_anonymous_variadic_dim`) are stored on persisted jaxtyping types (`cls.dtypes`, `cls.dims`). cloudpickle serialises them by value, so on the receiving side they are fresh `object()` instances that no longer match the live module-global sentinel — breaking `is`-checks that downstream shape-check code relies on. The symptom for `_anonymous_variadic_dim` is an `AssertionError` in `_check_dims`; the other two sentinels fail similarly when isinstance() runs against a cloudpickled type.

Replace each with a `__reduce__`-backed singleton class. The fourth `object()` sentinel in this file, `_not_made`, is transient factory output filtered before any type is built, so it doesn't suffer the same bug and is left alone.

Three regression tests added in `test_serialisation.py`, one per sentinel, each round-tripping under both `pickle` and `cloudpickle` and asserting `isinstance` after the round-trip — the call path that actually trips the bug under `cloudpickle`.

Bumps version 0.3.9 -> 0.3.10.

Fixes #390.

### CODE REVIEW

**Comment by patrick-kidger** [comment]
Thank you for the PR! Merged 🎉
